### DISCUSSION TITLE
Loop to swap data around in Data Processor?

### DISCUSSION THREAD

**Discussion by clemarescx** [WerWolv/ImHex#1608] [Get Help]
Hello,
thanks a lot for ImHex! Not sure if I'm having skill issues with the Data Processor - I'm new to node-based scripting - or if it's possible at all, but here goes:

I need to preprocess binary data as follow:
1. read data in 10 bytes chunks
2. reverse those 10 bytes
3. swap bytes 2 by 2

Example:
```
// 10-byte rows
00 01 02 03 04 05 06 07 08 09
0A 0B 0C 0D 0E 0F 10 11 12 13
```
should become
```
08 09 06 07 04 05 02 03 00 01
12 13 10 11 0E 0F 0C 0D 0A 0B
```

The equivalent in a pseudo C# code would be:
```csharp
byte[] data = someData;
const int rowLength = 10;
int rowsCount = data.Length / rowLength; // assume that the length is always evenly dividable by 10
for(int i = 0; i < rowsCount; i++)
{
    int offset = i * rowLength;
    byte[] row = data.Slice(offset, length: rowLength); // first row would be [ 01 02 03 04 ... ]
    row.Reverse();  // first row would now be [09 08 07 06 ...]
    for(int j = 0; j < row.Length; j += 2)
    {
        // swap bytes 2 by 2
        (row[j+1], row[j]) = (row[j], row[j+1]);
    }
    // first row would now be done: [08 09 06 07...]
    data.Write(offset, row, count: rowLength);
}
```
I see that there are node functions that I could use like Slice and Reverse, but I don't know how to look back to read from the Read buffer or a data buffer in chunks with an incremented offset... Or how to loop at all in general.
Any help would be appreciated.

**Comment by isti115** [comment]
I am trying to do something similar to this, and I see that a [Loop node](https://github.com/WerWolv/ImHex/blob/f3a9ca6d6f904befef3b72fdd1d008543804c85e/plugins/builtin/source/content/data_processor_nodes/control_nodes.cpp#L114-L148) has been [added since then](https://github.com/WerWolv/ImHex/commit/7975edade4046210f8ee732491ace236b4b871b3) by @WerWolv, but cannot figure out how to use it in any meaningful way. 🤔

**Reply by Lil-Nugs** [reply]
The code for the loop node just seems to output whatever is inputted every iteration? 

Or if you have start == end, it'll just output the initial value? What's the point of this node?

https://github.com/WerWolv/ImHex/blob/58228e0af4957229e74b554790d8cad4d032da95/plugins/builtin/source/content/data_processor_nodes/control_nodes.cpp#L123-L139
